### PR TITLE
Add metric for cpu user time usage of threads in the processing pool

### PR DESCRIPTION
**Pull request by drcrallen**
This is a small proposed patch to allow emitting of a user-cpu time statistic on metrics emitting thread pools (the processing pool).

The down side is that it no longer directly passes through the runnable or callable to the service, but instead wraps it in order to capture the actual cpu run time of the task.

### CODE REVIEW

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java [hunk 1]
```diff
@@ -51,17 +54,87 @@ protected ListeningExecutorService delegate()
...
+        } else {
+          threadUserTime = null;
+        }
+        final T retval = tCallable.call();
+        if (threadUserTime != null) {
+          emitter.emit(
+              metricBuilder
+                  .setDimension("priority", String.format("%d", getPriority()))
+                  .build("exec/cpu", threadMXBean.getCurrentThreadUserTime() - threadUserTime)
+          );
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: would be nice to have sys/user/wait as separate metrics...

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java [hunk 2]
```diff
@@ -51,17 +54,87 @@ protected ListeningExecutorService delegate()
...
+          threadUserTime = threadMXBean.getCurrentThreadUserTime();
+        } else {
+          threadUserTime = null;
+        }
+        final T retval = tCallable.call();
+        if (threadUserTime != null) {
+          emitter.emit(
+              metricBuilder
+                  .setDimension("priority", String.format("%d", getPriority()))
+                  .build("exec/cpu", threadMXBean.getCurrentThreadUserTime() - threadUserTime)
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: this will throw an NPE if threadUserTime is null

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: It's embedded in an if statement

Comment (fjy) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: can we call this segment/cpu or something similar?

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java [hunk 3]
```diff
@@ -51,17 +54,87 @@ protected ListeningExecutorService delegate()
...
+      }
+    };
     emitMetrics();
-    return delegate.submit(tCallable);
+    return delegate.submit(wrappedPrioritizedCallable);
   }
 
   @Override
-  public void execute(Runnable runnable)
+  public ListenableFuture<?> submit(final Runnable runnable)
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: why the change in method from execute to submit?

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java [hunk 4]
```diff
@@ -51,17 +54,87 @@ protected ListeningExecutorService delegate()
...
   {
+    final Callable<T> wrappedPrioritizedCallable = new PrioritizedCallable<T>()
+    {
+      @Override
+      public int getPriority()
+      {
+        if (tCallable instanceof PrioritizedCallable) {
+          return ((PrioritizedCallable) tCallable).getPriority();
+        } else {
+          return 0;
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: currently the default priority is specific to the PrioritizedExecutorService, so we should not make any assumptions here.

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java (hunk 2 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: we should probably add units too.

Comment (xvrl) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: @fjy do we have conventions on units as part of your metrics revamp?

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java [hunk 5]
```diff
@@ -51,17 +54,87 @@ protected ListeningExecutorService delegate()
...
+        if (threadMXBean.isThreadCpuTimeEnabled()) {
+          threadUserTime = threadMXBean.getCurrentThreadUserTime();
+        } else {
+          threadUserTime = null;
+        }
+        final T retval = tCallable.call();
+        if (threadUserTime != null) {
+          emitter.emit(
+              metricBuilder
+                  .setDimension("priority", String.format("%d", getPriority()))
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: No need for format here, Integer.toString() should be enough

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java (hunk 1 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: not sure if JMX provides that kind of information

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java (hunk 3 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: I made the two submits the ones with the bulk of the impl. the `execute` is still here, but down below and simply calls `submit`.

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java (hunk 2 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: Currently it reports nanoseconds. On my mac it only has resolution of microseconds.

This is currently used as the general processing pool, but not necessarily only segments. Unless we want the processing pool to be unique to segments I suggest keeping a more general name (but am open to suggestions on the matter)

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java (hunk 1 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: It provides User time and general CPU time. I can subtract the two to get "non user time" but didn't feel that was terribly useful since it couldn't be broken down anymore.

There are MANY types of cpu times reported by tools like `htop` which cannot be bubbled up through the jvm easily ( or consistently across host OS)

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java (hunk 5 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: fixed

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java (hunk 4 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: fixed

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java (hunk 2 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: @fjy Also note that `emitMetrics()` uses `exec/backlog`, so I was trying to be consistent in this file. It may make more sense to call it `processing/backlog` and `processing/cpu` ?

**Comment by drcrallen** [comment]
@xvrl  had suggested moving some of the prioritized stuff out of the MetricsEmittingExecutorService. I refactored things a bit so that wrappers are present in the MetricsEmittingExecutorService and PrioritizedExecutorService.

File: processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java [hunk 6]
```diff
@@ -23,14 +23,49 @@
...
+    {
+      return 0L;
+    }
+  };
+  private static final Runnable PRE = new Runnable()
+  {
+    @Override
+    public void run()
+    {
+      if (THREAD_MX_BEAN.isCurrentThreadCpuTimeSupported()) {
```
Comment (cheddar) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: instead of doing this if here, I think it would be more useful to initialize these runnables in a `static {}` initialization block that checks for support and, if supported, attaches `Runnable` objects that use it without question.  If not supported, it can build noop `Runnable` objects.

This is a micro-optimization, but, in general, any time you are asking a question of your system to figure out behavior, it's nicest to ask the question once and build objects that line up with the correct assumptions.  JIT can then do code-level optimization.

In this case, we would likely get the same performance gains out of branch prediction, but where it is easy, making the choice once and then just operating according to that choice should do better things.

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: I originally wasn't really sure at what level the bean would be consistent, as in: could you have some threads in a JVM that can support cpu time but not others? or could you have a flag enabled during runtime that could enable/disable CPUTime?

It looks like the Sun/Oracle way of doing things is to have the value assumed to not change per whole JVM, but I'm not sure if there's an official contract somewhere that enforces such a thing.

The proposed change makes sense though, and is pretty darn straight forward, so I'll put it in even though it is optimization without profiling.

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/MetricsEmittingExecutorService.java]: Fixed

**Comment by drcrallen** [comment]
Since this wasn't getting much support I tried #1696 to see if that approach would be better received.

**Comment by drcrallen** [comment]
Closing in favor of https://github.com/druid-io/druid/pull/1696
